ion-schema/src/constraint.rs

```diff
@@ -1031,6 +1067,117 @@ impl ConstraintValidator for FieldsConstraint {
     }
 }
 
+/// Implements an `field_names` constraint of Ion Schema
+/// [field_names]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#field_names
+#[derive(Debug, Clone, PartialEq)]
+pub struct FieldNamesConstraint {
+    type_reference: TypeReference,
+    requires_distinct: bool,
+}
+
+impl FieldNamesConstraint {
+    pub fn new(type_reference: TypeReference, requires_distinct: bool) -> Self {
+        Self {
+            type_reference,
+            requires_distinct,
+        }
+    }
+}
+
+impl ConstraintValidator for FieldNamesConstraint {
+    fn validate(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // create a set for checking duplicate elements
+        let mut element_set = vec![];
+
+        // get elements for given container in the form (ion_path_element, element_value)
+        let elements: Vec<(IonPathElement, &Element)> = match value {
+            IonSchemaElement::SingleElement(element) => {
+                // Check for null struct
+                if element.is_null() {
+                    return Err(Violation::new(
+                        "field_names",
+                        ViolationCode::TypeMismatched,
+                        "expected a struct but found null",
+                        ion_path,
+                    ));
+                }
+
+                return match element.as_struct() {
+                    Some(struct_value) => {
+                        let field_names: Vec<(IonPathElement, &Symbol)> = struct_value
```

Comment:
Given that the IonPathElement is the text of the &Symbol, is it necessary for this to be a `Vec<(IonPathElement, &Symbol)>`? My _opinion_ is that it would be easier to read/follow the code if we just stored `Vec<String>` and then constructed the `IonPathElement` in the loop that actually checks the field names.